pkg/bgpv1/manager/reconcilerv2/neighbor.go

```diff
@@ -0,0 +1,359 @@
+func (r *NeighborReconciler) Priority() int {
+	return 60
+}
+
+func (r *NeighborReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
+	if p.DesiredConfig == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil CiliumBGPNodeInstance")
+	}
+	if p.BGPInstance == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil BGPInstance")
+	}
+
+	var (
+		l = r.Logger.WithFields(logrus.Fields{
+			types.InstanceLogField: p.DesiredConfig.Name,
+		})
+
+		toCreate []*PeerData
+		toRemove []*PeerData
+		toUpdate []*PeerData
+		curNeigh []*PeerData = nil
+	)
+	newNeigh := p.DesiredConfig.Peers
+
+	l.Debug("Begin reconciling peers")
+
+	// get current configured peers
+	curInstance := r.getMetadata(p.BGPInstance)
+	if curInstance != nil {
+		curNeigh = curInstance[p.DesiredConfig.Name]
+	}
+
+	type member struct {
+		new *PeerData
+		cur *PeerData
+	}
+
+	nset := map[string]*member{}
+
+	for i, n := range newNeigh {
+		var (
+			key = n.Name
+			h   *member
+			ok  bool
+		)
+
+		config, err := r.getPeerConfig(n.PeerConfigRef.Name)
+		if err != nil {
+			return err
+		}
+
+		passwd, err := r.getPeerPassword(p.DesiredConfig.Name, n.Name, config)
+		if err != nil {
+			return err
+		}
+
+		if h, ok = nset[key]; !ok {
+			nset[key] = &member{
+				new: &PeerData{
+					Peer:     &newNeigh[i],
+					Config:   config,
+					Password: passwd,
+				},
+			}
+			continue
+		}
+		h.new = &PeerData{
+			Peer:     &newNeigh[i],
+			Config:   config,
+			Password: passwd,
+		}
+	}
+
+	for i, n := range curNeigh {
+		var (
+			key = n.Peer.Name
+			h   *member
+			ok  bool
+		)
+
+		if h, ok = nset[key]; !ok {
+			nset[key] = &member{
+				cur: curNeigh[i],
+			}
+			continue
+		}
+		h.cur = curNeigh[i]
+	}
+
+	for _, m := range nset {
+		// present in new neighbors (set new) but not in current neighbors (set cur)
+		if m.new != nil && m.cur == nil {
+			toCreate = append(toCreate, m.new)
+		}
+		// present in current neighbors (set cur) but not in new neighbors (set new)
+		if m.cur != nil && m.new == nil {
+			toRemove = append(toRemove, m.cur)
+		}
+		// present in both new neighbors (set new) and current neighbors (set cur), update if they are not equal
+		if m.cur != nil && m.new != nil {
+			if !m.cur.DeepEqual(m.new) {
+				toUpdate = append(toUpdate, m.new)
+			}
+		}
+	}
+
+	if len(toCreate) > 0 || len(toRemove) > 0 || len(toUpdate) > 0 {
+		l.Info("Reconciling peers for instance")
+	} else {
+		l.Debug("No peer changes necessary")
+	}
+
+	// create new neighbors
+	for _, n := range toCreate {
+		l.WithField(types.PeerLogField, n.Peer.Name).Info("Adding peer")
+
+		if err := p.BGPInstance.Router.AddNeighbor(ctx, types.NeighborRequest{
+			Peer:       n.Peer,
+			PeerConfig: n.Config,
+			Password:   n.Password,
+		}); err != nil {
+			return fmt.Errorf("failed to add neigbhor %s in instance %s: %w", n.Peer.Name, p.DesiredConfig.Name, err)
+		}
```

Comment:
We can revisit this later, but should we give up reconciliation immediately when `AddNeighbor` fails? Same for Update/Remove.

Reply:
I think we can revisit it later, it requires behavior change and more tests. Right now I wanted to keep behaviour similar to BGPv1.